File: assets/js/hooks/temperature_bands.js
import * as d3 from "d3"

const WIDTH = 820
const SHAPES_H = 130
const BANDS_H = 260
const MARGIN = { top: 18, right: 14, bottom: 26, left: 44 }

function triangular(x, a, b, c) {
  if (x <= a || x >= c) return 0
  if (x === b) return 1
  if (x < b) return b === a ? 1 : (x - a) / (b - a)
  return c === b ? 1 : (c - x) / (c - b)
}

function sampleTriangle(mf, bounds, steps = 80) {
  const [lo, hi] = bounds
  const out = []
  for (let i = 0; i <= steps; i++) {
    const x = lo + (i / steps) * (hi - lo)
    out.push([x, triangular(x, mf.a, mf.b, mf.c)])
  }
  return out
}

export const TemperatureBands = {
  mounted() {
    this.data = null
    this.el.style.position = "relative"

    this.handleEvent("update-bands", (data) => {
      this.data = data
      this.render()
    })

    this.pushEvent("sandbox:bands_ready", {})
  },

  render() {
    if (!this.data || !this.data.series || this.data.series.length === 0) return
    const { series, mfs, bounds } = this.data

    this.el.innerHTML = ""
    this.el.style.position = "relative"

    const container = d3.select(this.el)
      .append("div")
      .style("display", "flex")
      .style("flex-direction", "column")
      .style("gap", "14px")

    this.renderShapesPanel(container, mfs, bounds)
    this.renderBandsPanel(container, series, mfs)
  },

  renderShapesPanel(container, mfs, bounds) {
    const panel = container.append("div")

    panel.append("div")
      .attr("class", "text-[10px] uppercase tracking-[0.2em] text-gray-500 mb-1")
      .text("membership functions · drag sliders to reshape")

    const innerW = WIDTH - MARGIN.left - MARGIN.right
    const innerH = SHAPES_H - MARGIN.top - MARGIN.bottom

    const svg = panel.append("svg")
      .attr("viewBox", `0 0 ${WIDTH} ${SHAPES_H}`)
      .attr("preserveAspectRatio", "xMidYMid meet")
      .style("width", "100%")

    const g = svg.append("g")
      .attr("transform", `translate(${MARGIN.left},${MARGIN.top})`)

    const x = d3.scaleLinear().domain(bounds).range([0, innerW])
    const y = d3.scaleLinear().domain([0, 1]).range([innerH, 0])

    g.append("line")
      .attr("x1", 0).attr("x2", innerW)
      .attr("y1", innerH).attr("y2", innerH)
      .attr("stroke", "#374151")

    const line = d3.line().x(d => x(d[0])).y(d => y(d[1]))
    const area = d3.area()
      .x(d => x(d[0]))
      .y0(y(0))
      .y1(d => y(d[1]))

    mfs.forEach(mf => {
      const samples = sampleTriangle(mf, bounds)
      g.append("path")
        .attr("d", area(samples))
        .attr("fill", mf.color)
        .attr("fill-opacity", 0.18)

      g.append("path")
        .attr("d", line(samples))
        .attr("fill", "none")
        .attr("stroke", mf.color)
        .attr("stroke-width", 1.75)

      g.append("text")
        .attr("x", x(mf.b))
        .attr("y", y(1) - 5)
        .attr("fill", mf.color)
        .attr("font-size", 11)
        .attr("font-weight", "600")
        .attr("text-anchor", "middle")
        .text(mf.name)
    })

    g.append("g")
      .attr("transform", `translate(0,${innerH})`)
      .call(d3.axisBottom(x).ticks(9).tickFormat(d => `${d}°`))
      .call(sel => sel.selectAll("text").attr("fill", "#9ca3af").attr("font-size", 10))
      .call(sel => sel.selectAll("line").attr("stroke", "#374151"))
      .call(sel => sel.select(".domain").attr("stroke", "#374151"))
  },

  renderBandsPanel(container, series, mfs) {
    const panel = container.append("div")
      .style("position", "relative")

    panel.append("div")
      .attr("class", "text-[10px] uppercase tracking-[0.2em] text-gray-500 mb-1")
      .text("fuzzy memberships over time · hover for breakdown")

    const innerW = WIDTH - MARGIN.left - MARGIN.right
    const innerH = BANDS_H - MARGIN.top - MARGIN.bottom

    const svg = panel.append("svg")
      .attr("viewBox", `0 0 ${WIDTH} ${BANDS_H}`)
      .attr("preserveAspectRatio", "xMidYMid meet")
      .style("width", "100%")
      .style("display", "block")

    const g = svg.append("g")
      .attr("transform", `translate(${MARGIN.left},${MARGIN.top})`)

    const parseDate = d3.timeParse("%Y-%m-%d")
    const data = series.map(d => ({ date: parseDate(d.date), ...d.memberships }))
    const names = mfs.map(mf => mf.name)
    const colors = Object.fromEntries(mfs.map(mf => [mf.name, mf.color]))

    const x = d3.scaleTime()
      .domain(d3.extent(data, d => d.date))
      .range([0, innerW])

    const y = d3.scaleLinear()
      .domain([0, 1])
      .range([innerH, 0])

    const stacked = d3.stack()
      .keys(names)
      .value((d, key) => d[key] || 0)
      .offset(d3.stackOffsetNone)(data)

    const area = d3.area()
      .x(d => x(d.data.date))
      .y0(d => y(d[0]))
      .y1(d => y(d[1]))
      .curve(d3.curveMonotoneX)

    g.selectAll("path.band")
      .data(stacked)
      .join("path")
      .attr("class", "band")
      .attr("d", area)
      .attr("fill", d => colors[d.key] || "#666")
      .attr("opacity", 0.92)

    const [minDate, maxDate] = d3.extent(data, d => d.date)
    const years = d3.timeYear.range(
      d3.timeYear.floor(minDate),
      d3.timeYear.ceil(maxDate)
    )
    g.append("g")
      .attr("class", "year-ticks")
      .selectAll("line")
      .data(years)
      .join("line")
      .attr("x1", d => x(d))
      .attr("x2", d => x(d))
      .attr("y1", 0)
      .attr("y2", innerH)
      .attr("stroke", "#ffffff")
      .attr("stroke-opacity", 0.14)
      .attr("stroke-dasharray", "2,3")

    g.append("g")
      .attr("transform", `translate(0,${innerH})`)
      .call(d3.axisBottom(x).ticks(d3.timeYear.every(1)).tickFormat(d3.timeFormat("%Y")))
      .call(sel => sel.selectAll("text").attr("fill", "#9ca3af").attr("font-size", 10))
      .call(sel => sel.selectAll("line").attr("stroke", "#374151"))
      .call(sel => sel.select(".domain").attr("stroke", "#374151"))

    g.append("g")
      .call(d3.axisLeft(y).ticks(4).tickFormat(d3.format(".0%")))
      .call(sel => sel.selectAll("text").attr("fill", "#9ca3af").attr("font-size", 10))
      .call(sel => sel.selectAll("line").attr("stroke", "#374151"))
      .call(sel => sel.select(".domain").attr("stroke", "#374151"))

    stacked.forEach(s => {
      const lastPoint = s[s.length - 1]
      const mid = (lastPoint[0] + lastPoint[1]) / 2
      const bandHeight = Math.abs(y(lastPoint[1]) - y(lastPoint[0]))
      if (bandHeight < 11) return

      const base = g.append("text")
        .attr("x", innerW - 6)
        .attr("y", y(mid) + 3)
        .attr("font-size", 10)
        .attr("font-weight", "700")
        .attr("text-anchor", "end")
        .attr("paint-order", "stroke")
        .attr("stroke", "#0f172a")
        .attr("stroke-width", 3)
        .attr("stroke-linejoin", "round")
        .attr("fill", colors[s.key])
        .text(s.key)
      base.clone(true)
        .attr("stroke", "none")
    })

    const crosshair = g.append("line")
      .attr("y1", 0)
      .attr("y2", innerH)
      .attr("stroke", "#f8fafc")
      .attr("stroke-width", 1)
      .attr("opacity", 0)

    const tooltip = d3.select(this.el)
      .append("div")
      .style("position", "absolute")
      .style("pointer-events", "none")
      .style("background", "#0f172a")
      .style("border", "1px solid #1f2937")
      .style("border-radius", "6px")
      .style("padding", "8px 10px")
      .style("font-family", "ui-monospace, monospace")
      .style("font-size", "11px")
      .style("color", "#e5e7eb")
      .style("box-shadow", "0 4px 12px rgba(0,0,0,0.4)")
      .style("opacity", 0)
      .style("z-index", "20")

    const bisect = d3.bisector(d => d.date).left
    const hostEl = this.el
    const dateFmt = d3.timeFormat("%Y-%m-%d")

    g.append("rect")
      .attr("x", 0)
      .attr("y", 0)
      .attr("width", innerW)
      .attr("height", innerH)
      .attr("fill", "transparent")
      .on("mousemove", function (event) {
        const [mx] = d3.pointer(event, g.node())
        const d0 = x.invert(mx)
        const idx = Math.max(0, Math.min(data.length - 1, bisect(data, d0)))
        const d = data[idx]

        crosshair
          .attr("x1", x(d.date))
          .attr("x2", x(d.date))
          .attr("opacity", 0.85)

        const rows = names
          .map(name => {
            const pct = Math.round((d[name] || 0) * 100)
            return `<div style="display:flex;justify-content:space-between;gap:14px;">
              <span style="color:${colors[name]};">■ ${name}</span>
              <span style="color:#e5e7eb;">${pct}%</span>
            </div>`
          })
          .join("")

        const [sx, sy] = d3.pointer(event, hostEl)
        tooltip
          .html(
            `<div style="margin-bottom:4px;color:#9ca3af;">${dateFmt(d.date)}</div>${rows}`
          )
          .style("left", `${sx + 14}px`)
          .style("top", `${sy + 14}px`)
          .style("opacity", 1)
      })
      .on("mouseleave", () => {
        crosshair.attr("opacity", 0)
        tooltip.style("opacity", 0)
      })
  }
}

File: lib/fugue_web/live/menagerie_live/temperature_bands.ex
defmodule FugueWeb.MenagerieLive.TemperatureBands do
  @moduledoc """
  Server-rendered membership-function shapes + stacked-area band chart for the
  fuzzy temperature page. Replaces the former `TemperatureBands` JS hook.

  A tiny `BandsHover` hook handles crosshair + tooltip positioning on the
  bands panel; everything else is computed in Elixir.
  """

  use Phoenix.Component

  alias FugueWeb.MoodLive.SvgMath

  @width 820
  @shapes_h 130
  @bands_h 260
  @m_top 18
  @m_right 14
  @m_bottom 26
  @m_left 44
  @inner_w @width - @m_left - @m_right
  @shapes_inner_h @shapes_h - @m_top - @m_bottom
  @bands_inner_h @bands_h - @m_top - @m_bottom

  attr :series, :list, default: []
  attr :mfs, :list, default: []
  attr :shapes, :list, default: []
  attr :bounds, :list, default: [0.0, 48.0]

  def bands(assigns) do
    [lo, hi] = assigns.bounds
    temp_span = max(hi - lo, 1.0e-9)
    temp_x = fn t -> (t - lo) / temp_span * @inner_w end
    temp_y = fn v -> @shapes_inner_h - v * @shapes_inner_h end

    shapes_rendered =
      Enum.map(assigns.shapes, fn shape ->
        samples = shape.samples
        pts = Enum.map(samples, fn [t, v] -> {temp_x.(t), temp_y.(v)} end)

        line_d =
          case pts do
            [] ->
              ""

            [{x, y} | rest] ->
              "M#{SvgMath.fmt(x)},#{SvgMath.fmt(y)}" <>
                Enum.map_join(rest, "", fn {x, y} -> "L#{SvgMath.fmt(x)},#{SvgMath.fmt(y)}" end)
          end

        area_d =
          case pts do
            [] ->
              ""

            [{x0, _y0} | _] = list ->
              {xn, _} = List.last(list)
              base_y = SvgMath.fmt(temp_y.(0))

              "M#{SvgMath.fmt(x0)},#{base_y}L" <>
                Enum.map_join(list, "L", fn {x, y} -> "#{SvgMath.fmt(x)},#{SvgMath.fmt(y)}" end) <>
                "L#{SvgMath.fmt(xn)},#{base_y}Z"
          end

        %{
          name: shape.name,
          color: shape.color,
          line_d: line_d,
          area_d: area_d,
          label_x: temp_x.(shape.peak),
          label_y: temp_y.(1) - 5
        }
      end)

    x_temp_ticks =
      for t <- 0..8,
          v = lo + t / 8 * temp_span,
          do: %{x: temp_x.(v), label: "#{round(v)}°"}

    # --- Bands panel ---
    series = assigns.series
    mf_names = Enum.map(assigns.mfs, & &1.name)
    mf_colors = Map.new(assigns.mfs, &{&1.name, &1.color})

    {layers, year_lines, date_ticks, band_labels, series_json} =
      if series == [] do
        {[], [], [], [], "[]"}
      else
        dates = Enum.map(series, & &1.date)
        {min_d, max_d} = date_range(dates)
        span_days = max(Date.diff(max_d, min_d), 1)

        band_x = fn d -> Date.diff(Date.from_iso8601!(d), min_d) / span_days * @inner_w end
        band_y = fn v -> @bands_inner_h - v * @bands_inner_h end

        stacked =
          Enum.map(series, fn point ->
            {slices, _} =
              Enum.reduce(mf_names, {[], 0.0}, fn name, {acc, accumulated} ->
                val = Map.get(point.memberships, name, 0) * 1.0
                new_top = accumulated + val
                {[{name, accumulated, new_top} | acc], new_top}
              end)

            %{
              x: band_x.(point.date),
              slices: Enum.reverse(slices),
              date: point.date,
              memberships: point.memberships
            }
          end)

        layers =
          Enum.map(mf_names, fn name ->
            upper =
              Enum.map(stacked, fn p ->
                {_, _, y1} = List.keyfind(p.slices, name, 0)
                {p.x, band_y.(y1)}
              end)

            lower =
              Enum.map(stacked, fn p ->
                {_, y0, _} = List.keyfind(p.slices, name, 0)
                {p.x, band_y.(y0)}
              end)

            %{name: name, color: Map.get(mf_colors, name, "#666"), d: area_path(upper, lower)}
          end)

        year_lines =
          for y <- min_d.year..(max_d.year + 1),
              jan1 = Date.new!(y, 1, 1),
              Date.compare(jan1, min_d) != :lt and Date.compare(jan1, max_d) != :gt do
            %{x: band_x.(Date.to_iso8601(jan1))}
          end

        date_ticks =
          for y <- min_d.year..max_d.year,
              jan1 = Date.new!(y, 1, 1),
              Date.compare(jan1, min_d) != :lt and Date.compare(jan1, max_d) != :gt do
            %{x: band_x.(Date.to_iso8601(jan1)), label: Integer.to_string(y)}
          end

        band_labels =
          case List.last(stacked) do
            nil ->
              []

            last ->
              Enum.map(last.slices, fn {name, y0, y1} ->
                height = abs(band_y.(y1) - band_y.(y0))

                %{
                  name: name,
                  y: band_y.((y0 + y1) / 2) + 3,
                  color: Map.get(mf_colors, name, "#666"),
                  visible?: height >= 11
                }
              end)
          end

        # Per-day data for the hover tooltip, one row per day.
        # Shipped as JSON in a data attr; tiny JS hook bisects by x position.
        series_json =
          stacked
          |> Enum.map(fn p -> %{x: p.x, date: p.date, mems: p.memberships} end)
          |> Jason.encode!()

        {layers, year_lines, date_ticks, band_labels, series_json}
      end

    y_ticks =
      for t <- 0..4,
          v = t / 4,
          do: %{y: @bands_inner_h - v * @bands_inner_h, label: "#{round(v * 100)}%"}

    mfs_json = Jason.encode!(Enum.map(assigns.mfs, &Map.take(&1, [:name, :color])))

    assigns =
      assign(assigns,
        shapes_rendered: shapes_rendered,
        x_temp_ticks: x_temp_ticks,
        layers: layers,
        year_lines: year_lines,
        date_ticks: date_ticks,
        band_labels: band_labels,
        y_ticks: y_ticks,
        series_json: series_json,
        mfs_json: mfs_json,
        svg_width: @width,
        shapes_height: @shapes_h,
        bands_height: @bands_h,
        inner_w: @inner_w,
        shapes_inner_h: @shapes_inner_h,
        bands_inner_h: @bands_inner_h,
        g_transform: "translate(#{@m_left},#{@m_top})",
        shapes_axis_transform: "translate(0,#{@shapes_inner_h})",
        bands_axis_transform: "translate(0,#{@bands_inner_h})",
        m_left: @m_left,
        m_top: @m_top
      )

    ~H"""
    <div id="temperature-bands" phx-hook="BandsHover" style="position: relative;">
      <div style="display: flex; flex-direction: column; gap: 14px;">
        <div>
          <div class="text-[10px] uppercase tracking-[0.2em] text-gray-500 mb-1">
            membership functions · drag sliders to reshape
          </div>
          <svg
            viewBox={"0 0 #{@svg_width} #{@shapes_height}"}
            preserveAspectRatio="xMidYMid meet"
            style="width: 100%;"
          >
            <g transform={@g_transform}>
              <line x1="0" x2={@inner_w} y1={@shapes_inner_h} y2={@shapes_inner_h} stroke="#374151" />

              <%= for shape <- @shapes_rendered do %>
                <path d={shape.area_d} fill={shape.color} fill-opacity="0.18" />
                <path d={shape.line_d} fill="none" stroke={shape.color} stroke-width="1.75" />
                <text
                  x={SvgMath.fmt(shape.label_x)}
                  y={SvgMath.fmt(shape.label_y)}
                  fill={shape.color}
                  font-size="11"
                  font-weight="600"
                  text-anchor="middle"
                >
                  {shape.name}
                </text>
              <% end %>

              <g transform={@shapes_axis_transform}>
                <line x1="0" x2={@inner_w} y1="0" y2="0" stroke="#374151" />
                <%= for t <- @x_temp_ticks do %>
                  <line x1={SvgMath.fmt(t.x)} x2={SvgMath.fmt(t.x)} y1="0" y2="3" stroke="#374151" />
                  <text x={SvgMath.fmt(t.x)} y="14" text-anchor="middle" fill="#9ca3af" font-size="10">
                    {t.label}
                  </text>
                <% end %>
              </g>
            </g>
          </svg>
        </div>

        <div style="position: relative;">
          <div class="text-[10px] uppercase tracking-[0.2em] text-gray-500 mb-1">
            fuzzy memberships over time · hover for breakdown
          </div>
          <svg
            id="bands-svg"
            viewBox={"0 0 #{@svg_width} #{@bands_height}"}
            preserveAspectRatio="xMidYMid meet"
            style="width: 100%; display: block;"
            data-series={@series_json}
            data-mfs={@mfs_json}
            data-inner-w={@inner_w}
            data-inner-h={@bands_inner_h}
            data-m-left={@m_left}
            data-m-top={@m_top}
          >
            <g transform={@g_transform}>
              <%= for layer <- @layers do %>
                <path class="band" d={layer.d} fill={layer.color} opacity="0.92" />
              <% end %>

              <g class="year-ticks">
                <%= for y <- @year_lines do %>
                  <line
                    x1={SvgMath.fmt(y.x)}
                    x2={SvgMath.fmt(y.x)}
                    y1="0"
                    y2={@bands_inner_h}
                    stroke="#ffffff"
                    stroke-opacity="0.14"
                    stroke-dasharray="2,3"
                  />
                <% end %>
              </g>

              <g transform={@bands_axis_transform}>
                <line x1="0" x2={@inner_w} y1="0" y2="0" stroke="#374151" />
                <%= for t <- @date_ticks do %>
                  <line x1={SvgMath.fmt(t.x)} x2={SvgMath.fmt(t.x)} y1="0" y2="3" stroke="#374151" />
                  <text x={SvgMath.fmt(t.x)} y="16" text-anchor="middle" fill="#9ca3af" font-size="10">
                    {t.label}
                  </text>
                <% end %>
              </g>

              <g>
                <line x1="0" x2="0" y1="0" y2={@bands_inner_h} stroke="#374151" />
                <%= for t <- @y_ticks do %>
                  <line x1="-3" x2="0" y1={SvgMath.fmt(t.y)} y2={SvgMath.fmt(t.y)} stroke="#374151" />
                  <text
                    x="-6"
                    y={SvgMath.fmt(t.y + 3)}
                    text-anchor="end"
                    fill="#9ca3af"
                    font-size="10"
                  >
                    {t.label}
                  </text>
                <% end %>
              </g>

              <%= for label <- @band_labels, label.visible? do %>
                <text
                  x={@inner_w - 6}
                  y={SvgMath.fmt(label.y)}
                  font-size="10"
                  font-weight="700"
                  text-anchor="end"
                  paint-order="stroke"
                  stroke="#0f172a"
                  stroke-width="3"
                  stroke-linejoin="round"
                  fill={label.color}
                >
                  {label.name}
                </text>
              <% end %>

              <line
                class="bands-crosshair"
                y1="0"
                y2={@bands_inner_h}
                stroke="#f8fafc"
                stroke-width="1"
                opacity="0"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
    """
  end

  defp area_path(upper, lower) do
    upper_path = SvgMath.basis_path(upper)
    lower_path = SvgMath.basis_path(Enum.reverse(lower))

    lower_continued =
      case lower_path do
        "M" <> rest -> "L" <> rest
        "" -> ""
        other -> other
      end

    upper_path <> lower_continued <> "Z"
  end

  defp date_range([first | _] = dates) do
    parsed = Enum.map(dates, &Date.from_iso8601!/1)
    {Enum.min(parsed, Date), Enum.max(parsed, Date)}
  rescue
    _ -> {Date.from_iso8601!(first), Date.from_iso8601!(first)}
  end
end
